### PR TITLE
Add prescribed viscosity plugin.

### PR DESCRIPTION
**Pull request by MFraters**
For a project I am working on we need to prescribe a viscosity at a location, which moves with time in a very similar way we can do with the prescribed velocity cookbook plugin. We first hard-coded it in a material model. After a discussion with @jdannberg, @gassmoeller, @alarshi and @RanpengLi, we think this may have broader use. Personally I think this would have been nice to have in previous projects. So we made it into a plugin which can be used as use other material models as a base model. 

The code is basically a combination of the depth dependent plugin and the prescribed velocity plugin.

I do need to finish cleaning the code (see todo's and missing/unupdated documentation and missing changelog entry) and making tests, but I wanted to have this out here early to make sure that this plugin is both wanted and that the naming of the plugin and functions is fine with everyone. I have already tested it for the project I am working on, and it works as intended.

### CODE REVIEW

File: include/aspect/material_model/prescribed_viscosity.h [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+         * Parsed function that specifies viscosity depth-dependence when using the Function
+         * method.
+         */
+        Functions::ParsedFunction<dim> prescribed_viscosity_indicator_function;
+
+        /**
+         * Parsed function that specifies viscosity in the region described by
+         */
+        Functions::ParsedFunction<dim> prescribed_viscosity_function;
+
```
Comment (gassmoeller) [inline include/aspect/material_model/prescribed_viscosity.h:99]: one empty line

File: include/aspect/material_model/prescribed_viscosity.h [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    using namespace dealii;
+
+    /**
+     * A material model that applies a depth-dependent viscosity to a ''base model''
+     * chosen from any of the other available material models. This depth-dependent
+     * material model allows the user to specify a depth-dependent reference viscosity
+     * either through a parsed function, lists of depth and viscosity values, or a file.
+     * The current implementation only allows for depth-dependence of viscosity - all
+     * other properties are derived from the base model.
+     * @ingroup MaterialModels // TODO
```
Comment (gassmoeller) [inline include/aspect/material_model/prescribed_viscosity.h:45]: TODO?

File: include/aspect/material_model/prescribed_viscosity.h [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+#ifndef _aspect_material_model_depth_dependent_h
+#define _aspect_material_model_depth_dependent_h
+
+#include <deal.II/base/function_lib.h>
+#include <aspect/material_model/interface.h>
+#include <aspect/simulator_access.h>
+#include <deal.II/base/parsed_function.h>
+#include <aspect/material_model/rheology/ascii_depth_profile.h>
+
+#include <boost/lexical_cast.hpp>
```
Comment (gassmoeller) [inline include/aspect/material_model/prescribed_viscosity.h:30]: unnecessary include?

File: source/material_model/prescribed_viscosity.cc [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+*/
+
+#include <aspect/material_model/prescribed_viscosity.h>
+#include <aspect/utilities.h>
+#include <aspect/geometry_model/interface.h>
+
+#include <array>
+
+#include <utility>
+#include <limits>
```
Comment (gassmoeller) [inline source/material_model/prescribed_viscosity.cc:28]: I dont think you need many of these includes?

File: source/material_model/prescribed_viscosity.cc [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+          if (in.requests_property(MaterialProperties::viscosity))
+            {
+              const double indicator = prescribed_viscosity_indicator_function.value(in.position[i]);
+              const double viscosity = prescribed_viscosity_function.value(in.position[i]);
+
+              if (indicator > 0.5)
+                {
+                  out.viscosities[i] = viscosity;
+                }
+            }
```
Comment (gassmoeller) [inline source/material_model/prescribed_viscosity.cc:83]: Maybe like this (saves a tiny amount of time):
```suggestion
          if (in.requests_property(MaterialProperties::viscosity))
            {
              const double indicator = prescribed_viscosity_indicator_function.value(in.position[i]);

              if (indicator > 0.5)
                {
                  out.viscosities[i] = prescribed_viscosity_function.value(in.position[i]);
                }
            }
```

File: source/material_model/prescribed_viscosity.cc [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+        prm.enter_subsection ("Prescribed viscosity");
+        {
+
+          prm.declare_entry("Base model","simple",
+                            Patterns::Selection(MaterialModel::get_valid_model_names_pattern<dim>()),
+                            "The name of a material model that will be modified by a depth "
+                            "dependent viscosity. Valid values for this parameter "
+                            "are the names of models that are also valid for the "
+                            "``Material models/Model name'' parameter. See the documentation for "
+                            "that for more information.");
```
Comment (gassmoeller) [inline source/material_model/prescribed_viscosity.cc:104]: Update description.

**Review by gassmoeller — COMMENTED** [review]
Yes, looks good and thanks for setting it up. I left some comments.
What we would need in addition:
- a test
- a changelog entry.

**Comment by MFraters** [comment]
Thanks for the review! I have addressed your comments and added a test and a changelog.

I made sure that the test is happy, and I hope the tester is as well :)

happy_viscosity_plugin

File: source/material_model/prescribed_viscosity.cc [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+                throw;
+              }
+          }
+          prm.leave_subsection();
+        }
+        prm.leave_subsection ();
+      }
+      prm.leave_subsection ();
+
+      /* After parsing the parameters for depth dependent, it is essential to parse
```
Comment (gassmoeller) [inline source/material_model/prescribed_viscosity.cc:166]: depth dependent

File: source/material_model/prescribed_viscosity.cc [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+          {
+            Functions::ParsedFunction<dim>::declare_parameters (prm, 3);
+          }
+          prm.leave_subsection ();
+        }
+        prm.leave_subsection ();
+      }
+      prm.leave_subsection ();
+    }
+
```
Comment (gassmoeller) [inline source/material_model/prescribed_viscosity.cc:108]: three empty lines

File: source/material_model/prescribed_viscosity.cc [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+      this->model_dependence = base_model->get_model_dependence();
+    }
+
+    template <int dim>
+    bool
+    PrescribedViscosity<dim>::
+    is_compressible () const
+    {
+      return base_model->is_compressible();
+    }
```
Comment (gassmoeller) [inline source/material_model/prescribed_viscosity.cc:178]: Please move this function above the parse_parameters and declare_parameters functions. This is not a hard rule, but I would like to keep our convention in as many files as possible. Makes it easier to read.

File: source/material_model/prescribed_viscosity.cc [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+{
+  namespace MaterialModel
+  {
+    ASPECT_REGISTER_MATERIAL_MODEL(PrescribedViscosity,
+                                   "prescribed viscosity",
+                                   "A material model that applies a viscosity to a ''base model'' chosen from any of "
+                                   "the other available material models. This prescribed viscosity material model "
+                                   "allows the user to specify a function which describes where the viscosity should be "
+                                   "prescribed and a second function which describes the viscosity in that region. "
+                                   "This material model requires a base model which prescribes the viscosity in the "
```
Comment (gassmoeller) [inline source/material_model/prescribed_viscosity.cc:193]: ... and the other material parameters ...

File: tests/happy_prescribed_viscosity_plugin.prm
Comment (gassmoeller) [inline tests/happy_prescribed_viscosity_plugin.prm:1]: maybe rename to `prescribed_viscosity_happy.prm`, this will be easier to find and understand

**Review by gassmoeller — COMMENTED** [review]
Thanks for the morale booster :-)
Some comments, but afterwards this should be ready. The failing tests on deal.II master are unrelated.

**Comment by MFraters** [comment]
Thanks for the review, I have addressed your comments! :)

**Review by gassmoeller — APPROVED** [review]
Thank you!

**Pull request merged**